RFS/src/main/java/com/rfs/worker/IndexStep.java

```diff
@@ -0,0 +1,475 @@
+            return cmsEntry.orElseThrow(
+                () -> new MissingIndexEntry()
+            );
+        }
+    }
+
+    public static abstract class Base implements WorkerStep {
+        protected final Logger logger = LogManager.getLogger(getClass());
+        protected final SharedMembers members;
+    
+        public Base(SharedMembers members) {
+            this.members = members;
+        }
+    }
+
+    /*
+     * Updates the Worker's phase to indicate we're doing work on an Index Migration
+     */
+    public static class EnterPhase extends Base {
+        public EnterPhase(SharedMembers members) {
+            super(members);
+        }
+
+        @Override
+        public void run() {
+            logger.info("Index Migration not yet completed, entering Index Phase...");
+            members.globalState.updatePhase(GlobalState.Phase.INDEX_IN_PROGRESS);
+        }
+
+        @Override
+        public WorkerStep nextStep() {
+            return new GetEntry(members);
+        }
+    }
+
+    /*
+     * Gets the current Index Migration entry from the CMS, if it exists
+     */
+    public static class GetEntry extends Base {
+
+        public GetEntry(SharedMembers members) {
+            super(members);
+        }
+
+        @Override
+        public void run() {
+            logger.info("Pulling the Index Migration entry from the CMS, if it exists...");
+            members.cmsEntry = members.cmsClient.getIndexEntry();
+        }
+
+        @Override
+        public WorkerStep nextStep() {
+            if (members.cmsEntry.isEmpty()) {
+                return new CreateEntry(members);
+            } 
+            
+            CmsEntry.Index currentEntry = members.cmsEntry.get();            
+            switch (currentEntry.status) {
+                case SETUP:
+                    // TODO: This uses the client-side clock to evaluate the lease expiration, when we should
+                    // ideally be using the server-side clock.  Consider this a temporary solution until we find
+                    // out how to use the server-side clock.
+                    long leaseExpiryMillis = Long.parseLong(currentEntry.leaseExpiry);
+                    Instant leaseExpiryInstant = Instant.ofEpochMilli(leaseExpiryMillis);
+                    boolean leaseExpired = leaseExpiryInstant.isBefore(Instant.now());
+
+                    // Don't try to acquire the lease if we're already at the max number of attempts
+                    if (currentEntry.numAttempts >= CmsEntry.Index.MAX_ATTEMPTS && leaseExpired) {
+                        return new ExitPhaseFailed(members, new MaxAttemptsExceeded());
+                    }
+
+                    if (leaseExpired) {
+                        return new AcquireLease(members);
+                    } 
+                    
+                    logger.info("Index Migration entry found, but there's already a valid work lease on it");
+                    return new RandomWait(members);
```

Comment:
Per discussion, this is when you find that the one possible thing for you to work on (setting up the index work items) isn't available to you because someone else has a lease on it.  We do a random backoff to prevent all the RFS Workers slamming the Target Cluster as fast as they can to see if the lease has expired.